core: optimize timers, conn handler and responses for throughput under high concurrency

Bogdanp opened the pull request:
This change improves the web server's throughput with many (1k+) concurrent connections.  Of the bunch, the most important changes are those to the timer manager and the change to stop peeking in `handle-connection/cm`. I'd recommend looking at the commits individually for readability.

The change to `add-missing-headers` also fixes an issue: previously, the check if a header was already set on the response was case-sensitive.

On my local machine (macOS, 2.4GHz), [this app](https://gist.github.com/Bogdanp/b5ce64066fe6093da7e2ef70c3dab012) performs as follows when 4k requests per second are made for a minute after it's been warmed up:

* master:

```
$ cat targets | vegeta attack -duration=60s -rate 4000/s | vegeta report
Requests      [total, rate, throughput]         240000, 4000.02, 3999.85
Duration      [total, attack, wait]             1m0s, 1m0s, 2.585ms
Latencies     [min, mean, 50, 90, 95, 99, max]  542.112µs, 3.733ms, 3.649ms, 4.776ms, 5.147ms, 5.946ms, 70.843ms
Bytes In      [total, mean]                     1440000, 6.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:240000
Error Set:
```

* this branch:

```
$ cat targets | vegeta attack -duration=60s -rate 4000/s | vegeta report
Requests      [total, rate, throughput]         240000, 4000.02, 4000.00
Duration      [total, attack, wait]             1m0s, 1m0s, 288.141µs
Latencies     [min, mean, 50, 90, 95, 99, max]  174.083µs, 295.654µs, 271.062µs, 381.418µs, 462.48µs, 646.391µs, 5.472ms
Bytes In      [total, mean]                     1440000, 6.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:240000
Error Set:
```

On [one of my servers](https://www.hetzner.com/dedicated-rootserver/ax61-nvme), running the [TechEmpower Benchmarks](https://www.techempower.com/benchmarks) against the two branches (`racket` is running master and `racket-perf` is running this branch) yields:

https://www.techempower.com/benchmarks/#section=test&shareid=669bfab7-9242-4c26-8921-a4fe9ccd8530&hw=ph&test=composite&a=2

The maximum throughput does not increase by much (only about 4k/s in the json and plaintext benchmarks), but take a look at the `Data Table` tab for each test, in particular the plaintext test. Performance now suffers less when there are many concurrent connections.

There are [more changes that could be made](https://gist.githubusercontent.com/Bogdanp/738ea20d7454f527eeb154b03fd9768b/raw/b045f308c590d86530ef6b00ee4082b74bc82948/perf-results.txt), but this is all I had time for this weekend and I think these are some nice gains.

jeapostrophe commented on web-server-lib/web-server/private/timer.rkt:
```diff
@@ -18,73 +19,84 @@
-;; Limitation on this add-timer: thunk cannot make timer
-;;  requests directly, because it's executed directly by
-;;  the timer-manager thread
+  (define ch (make-async-channel))
```
I think it should be changed to use the thread's mailbox. This code was written before thread mailboxes and we don't need to the async channel to be separate for the servicing thread.

Bogdanp commented:
Thanks for the comments! I'll make changes to address them this weekend.

Bogdanp commented on web-server-lib/web-server/http/response.rkt:
```diff
@@ -348,17 +392,18 @@
                       [(list)
                        ; Final boundary (must start on new line; ends with a new line)
-                       (fprintf (connection-o-port conn) "--~a--\r\n" boundary)
+                       (cprintf (connection-o-port conn) #"--~a--\r\n" boundary)
```
A boundary is always generated in `output-file` (line 306) when there is more than one range. The `then` clause of the `if` surrounding this loop handles the case where there is a single range and there are never 0 ranges: if the request doesn't specify a list of ranges, then a single range for the entire file is created at line 346.

Bogdanp commented on web-server-lib/web-server/http/response.rkt:
```diff
@@ -163,14 +207,13 @@
         ;; *something* every (current-send-timeout) seconds.
         (reset-connection-response-send-timeout! conn)
-        (fprintf to-client "~a\r\n" (number->string bytes-read-or-eof 16))
+        (cprintf to-client #"~a\r\n" (number->string bytes-read-or-eof 16))
```
In the standard library? I think it would be. I feel like I've had to write something like it in a few places.

Bogdanp commented on web-server-lib/web-server/http/response.rkt:
```diff
@@ -54,53 +55,44 @@
-         [#"Connection" #"close"])
-        empty)
-    hs)))
+(define-syntax (add-missing-headers stx)
```
It's used to add the headers iff they don't exist so I think the checking is necessary since the app could provide its own versions of these headers.

Bogdanp commented:
I've addressed all the comments. The numbers now look like this:

```
Requests      [total, rate, throughput]         240000, 4000.03, 4000.01
Duration      [total, attack, wait]             1m0s, 1m0s, 265.431µs
Latencies     [min, mean, 50, 90, 95, 99, max]  168.712µs, 261.615µs, 248.794µs, 309.237µs, 354.592µs, 541.572µs, 3.149ms
Bytes In      [total, mean]                     1440000, 6.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:240000
Error Set:
```

samth commented on web-server-lib/web-server/private/timer.rkt:
```diff
@@ -18,73 +18,84 @@
-  #:mutable)
+(define-logger web-server/timer)
+
+(struct timer-manager (thd))
```
Is this struct necessary now?

Bogdanp replied:
No, but I kept it to minimize the changes since it probably doesn't add much overhead (relative to other things). I'll remove it if necessary when I do some more profiling.

samth commented on web-server-lib/web-server/private/timer.rkt:
```diff
@@ -18,73 +18,84 @@
+         (log-web-server/timer-debug "timers: ~a, next deadline in ~.sms" (hash-count timers) (- first-deadline (current-inexact-milliseconds)))
+         (sync
+          (handle-evt
+           (thread-receive-evt)
```
It would be nice if this produced the message, instead of the event, so that you didn't have to deal with the thread system twice in this code.

Bogdanp replied:
Yeah, I've written enough code like this now that I'm used to it, but the behavior of `thread-receive-evt` was definitely surprising initially. Looking at `thread/thread.rkt`, it wouldn't be hard to add a variant that dequeues a message, but I haven't been able to come up with a good name for it. `thread-dequeue-evt` seems like it would cause confusion.

jeapostrophe commented:
This looks ready to me; are there any other plans?

Bogdanp commented:
I plan to do some more profiling in the coming weeks but I don't know exactly when that will be. In the mean time, I think these can be merged.

jeapostrophe commented:
Awesome!
